### PR TITLE
add tikv client

### PR DESCRIPTION
**Pull request by morefreeze**
@disksing PTAL

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	return &Client{
+		dst:   *addr,
+		msgID: 0,
+		pool:  p,
+	}
+}
+
+// Send send a Request and receive Response.
+func (c *Client) Send(req *pb.Request) (*pb.Response, error) {
+	conn, err := c.pool.Get()
```
Comment (disksing) [inline client.go]: What's the behaviour  of `pool.Get()` if all connections are in use?

File: client_test.go [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+	}
+	log.Debug("Start listenning...")
+	go func() {
+		for {
+			conn, err := s.l.Accept()
+			if err != nil {
+				log.Fatal("Error accepting: ", err)
+			}
+			go handleRequest(conn)
+		}
```
Comment (disksing) [inline client_test.go]: We may need a `TearDown` to stop the loop.

Comment (disksing) [inline client_test.go]: Oh, already got one, my mistake. But is it safe to access the Listener with out lock?

File: client.go (hunk 1 above)
Comment (morefreeze) [inline client.go]: It still call `factory` if pool is empty. But pool only keep maxConneciton when conn.Close(), redundant connection will be dropped.

Comment (disksing) [inline client.go]: Got it.

**Comment by disksing** [comment]
LGTM

**Comment by morefreeze** [comment]
@siddontang PTAL

File: client.go (hunk 1 above)
Comment (siddontang) [inline client.go]: should check conn is still alive.

File: client.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer conn.Close()
+	c.mu.Lock()
+	c.msgID++
```
Comment (siddontang) [inline client.go]: why not using atomic add instead?

File: client_test.go [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+	s.l, err = net.Listen("tcp", fmt.Sprintf("%s:%d", ip, port))
+	if err != nil {
+		log.Fatal("Error listening: ", err)
+	}
+	log.Debug("Start listenning...")
+	go func() {
+		for {
+			conn, err := s.l.Accept()
+			if err != nil {
+				log.Fatal("Error accepting: ", err)
```
Comment (siddontang) [inline client_test.go]: can't fatal. 
we close the listener in teardown, so here accept will raise an error. use log.Error instead.

File: kv.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	mc.cache[uuid] = s
+	return s, nil
+}
+
+type tikvStore struct {
+	mu        sync.Mutex
+	uuid      string
+	storeName string
+	oracle    oracle.Oracle
+	conn      *Client
```
Comment (siddontang) [inline kv.go]: conn -> client

File: kv.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+func (s *tikvStore) getTiKVClient() *Client {
```
Comment (siddontang) [inline kv.go]: getClient is ok.

File: client_test.go (hunk 4 above)
Comment (morefreeze) [inline client_test.go]: Got it

File: client.go (hunk 3 above)
Comment (morefreeze) [inline client.go]: But I need assign `msgID` to `curMsgID` atomically.

Comment (morefreeze) [inline client.go]: But I need assign `msgID` to `curMsgID` atomically.

**Comment by siddontang** [comment]
Rest LGTM

File: client.go (hunk 1 above)
Comment (morefreeze) [inline client.go]: Got it

**Comment by morefreeze** [comment]
@siddontang PTAL

File: client.go [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+}
+
+// Close close client
+func (c *Client) Close() error {
+	c.pool.Close()
+	return nil
+}
+
+func (c *Client) doSend(conn net.Conn, req *pb.Request) (*pb.Response, error) {
+	defer conn.Close()
```
Comment (siddontang) [inline client.go]: move defer after doSend above.

File: client_test.go [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+	s.l, err = net.Listen("tcp", fmt.Sprintf("%s:%d", ip, port))
+	if err != nil {
+		log.Fatal("Error listening: ", err)
+	}
+	log.Debug("Start listenning...")
+	go func() {
+		for {
+			conn, err := s.l.Accept()
+			if err != nil {
+				log.Error("Error accepting: ", err)
```
Comment (siddontang) [inline client_test.go]: return the loop.

File: kv.go [hunk 9]
```diff
@@ -0,0 +1,161 @@
...
+}
+
+func (s *tikvStore) Close() error {
+	mc.mu.Lock()
+	defer mc.mu.Unlock()
+
+	delete(mc.cache, s.uuid)
+
+	var err error
+	err = s.client.Close()
```
Comment (siddontang) [inline kv.go]: err := s.client.Close()

File: snapshot.go [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+	pb "github.com/pingcap/ticlient/kvrpc/proto"
+	"github.com/pingcap/tidb/kv"
+)
+
+var (
+	_ kv.Snapshot = (*tikvSnapshot)(nil)
+	_ kv.Iterator = (*tikvIter)(nil)
+)
+
+const tikvBatchSize = 1000
```
Comment (siddontang) [inline snapshot.go]: const (

)

File: snapshot.go [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+		CmdScanReq: cmdScanReq,
+	}
+
+	log.Debugf("start_key %d %s", len(k), k)
+	resp, err := s.store.client.Send(req)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if resp.GetType() != pb.MessageType_CmdScan {
+		return nil, errors.Errorf("receive response[%s] dismatch request[%s]",
```
Comment (siddontang) [inline snapshot.go]: dismatches

File: client.go [hunk 12]
```diff
@@ -0,0 +1,114 @@
...
+	curMsgID := atomic.AddUint64(&c.msgID, 1)
+	log.Debugf("Send request msgID[%d] type[%s]", curMsgID, req.GetType())
+	err := kvrpc.EncodeMessage(conn, curMsgID, req)
+	if err != nil {
+		// This connection is not valid any more, so close its underlying conn.
+		if connPool, ok := conn.(*pool.PoolConn); ok {
```
Comment (siddontang) [inline client.go]: for every error, there is no need to reuse this connection too. 
so you can use it after doSend failed.

**Comment by morefreeze** [comment]
@siddontang PTAL

**Comment by siddontang** [comment]
Rest LGTM

**Pull request merged**